### PR TITLE
Remove `spill_lock` parameter from `SpillableBuffer.get_ptr()`

### PR DESCRIPTION
**Pull request by madsbk**
This simplifies the code and is a step towards `copy-on-write`: https://github.com/rapidsai/cudf/pull/11718#issuecomment-1382361982

cc. @galipremsagar, @shwina, @vyasr 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by galipremsagar — APPROVED** [review]

File: python/cudf/cudf/_lib/column.pyx [hunk 1]
```diff
@@ -533,12 +527,9 @@ cdef class Column:
...
                 isinstance(data_owner, SpillableBuffer) and
                 # We check that `data_owner` is spill locked (not spillable)
-                # and that its pointer is the same as `data_ptr` _without_
-                # exposing the buffer permanently (calling get_ptr with a
-                # dummy SpillLock).
+                # and that it points to the same memory as `data_ptr`.
                 not data_owner.spillable and
-                data_owner.get_ptr(spill_lock=SpillLock()) == data_ptr and
-                data_owner.size == base_nbytes
+                data_owner.memory_info() == (data_ptr, base_nbytes, "gpu")
```
Comment (shwina) [inline python/cudf/cudf/_lib/column.pyx:532]: Would this not result in the data owner being permanently unspillable?

Comment (shwina) [reply python/cudf/cudf/_lib/column.pyx:532]: Never mind -- i see that this is not the case

File: python/cudf/cudf/core/buffer/spillable_buffer.py [hunk 2]
```diff
@@ -270,35 +265,70 @@ def spill_lock(self, spill_lock: SpillLock) -> None:
...
+        return self._ptr
 
-        self.spill_lock(spill_lock)
-        self._last_accessed = time.monotonic()
+    def memory_info(self) -> Tuple[int, int, str]:
+        """Get pointer, size, and device type of this buffer.
+
+        Warning, it is not safe to access the pointer value without
+        spill lock the buffer manually. This method neither expose
+        nor spill lock the buffer.
```
Comment (shwina) [inline python/cudf/cudf/core/buffer/spillable_buffer.py:302]:
```suggestion
        spill lock the buffer manually. This method neither exposes
        nor spill locks the buffer.
```

**Review by shwina — APPROVED** [review]
@wence- could you also take a look at this when you have a moment?

**Comment by madsbk** [comment]
I can confirm that running the full test suite with spilling passes locally.
```python
CUDF_SPILL=on CUDF_SPILL_DEVICE_LIMIT=1 pytest python/cudf/cudf/tests/
```

**Review by wence- — APPROVED** [review]
Very minor typo nit, but otherwise looks good I think, thanks.

**Comment by wence-** [comment]
All the dask-cudf tests failed in one case https://github.com/rapidsai/cudf/actions/runs/3956159947/jobs/6775597195 Perhaps there was a duff GPU? Everything failed with device memory allocation errors.

File: python/cudf/cudf/core/buffer/spillable_buffer.py [hunk 3]
```diff
@@ -239,14 +239,10 @@ def spill(self, target: str = "cpu") -> None:
...
+        """Mark the buffer as "exposed" and make it unspillable permanently.
 
-        Notice, this will mark the buffer as "exposed" and make
-        it unspillable permanently.
-
-        Consider using `.get_ptr()` instead.
+        This also unspill the buffer (unspillable buffers cannot be spilled!).
         """
 
         self._manager.spill_to_device_limit()
```
[the remark this answers is not in the document]
Comment (madsbk) [reply python/cudf/cudf/core/buffer/spillable_buffer.py:248]: Good point but notice `spill_to_device_limit()` is mostly for testing and debugging. 
By default,  `CUDF_SPILL_DEVICE_LIMIT` isn't set, which makes `spill_to_device_limit()` a no-op.

I think in most real world scenarios, we should rely on spill-on-demand.

**Comment by galipremsagar** [comment]
Now I see an unrelated pytest failure: 
```
FAILED python/cudf/cudf/tests/test_parquet.py::test_parquet_writer_nested[data2]
```

It's a trivial fix where we might have to un-xfail it. Testing it locally to open a PR.

**Pull request merged**